ASoC: SOF: Intel: hda-pcm: Limit the maximum number of periods by MAX…

ujfalusi opened the pull request:
…_BDL_ENTRIES

The number of periods for HDA is limited by the maximum number of BDL entries we have space for (which is 255).

Make sure that we don't allow more periods than the maximum number of BDL entries minus one.

ujfalusi committed:
ASoC: SOF: Intel: hda-pcm: Limit the maximum number of periods by MAX_BDL_ENTRIES

The HDaudio specification Section 3.6.2 limits the number of BDL entries to 256.

Make sure we don't allow more periods than this normative value.

Signed-off-by: Peter Ujfalusi <peter.ujfalusi@linux.intel.com>

plbossart commented on sound/soc/sof/intel/hda-pcm.c:
```diff
@@ -258,6 +258,12 @@ int hda_dsp_pcm_open(struct snd_sof_dev *sdev,
 	snd_pcm_hw_constraint_integer(substream->runtime,
 				      SNDRV_PCM_HW_PARAM_PERIODS);
 
+	/* Limit the maximum number of periods to not exceed the BDL etries count */
```
typo: entries

plbossart commented on sound/soc/sof/intel/hda-pcm.c:
```diff
@@ -258,6 +258,12 @@ int hda_dsp_pcm_open(struct snd_sof_dev *sdev,
+		snd_pcm_hw_constraint_minmax(runtime, SNDRV_PCM_HW_PARAM_PERIODS,
+					     runtime->hw.periods_min,
+					     HDA_DSP_MAX_BDL_ENTRIES - 1);
+
```
The spec says up to 256 entries are legit, this should be MAX_BDL_ENTRIES, not a MAX_BDL_ENTRIES - 1


commit message suggested edit:

The HDaudio specification Section 3.6.2 limits the number of BDL entries to 256.

Make sure we don't allow more periods than this normative value.

ujfalusi commented:
Changes since v1:
- fixed off by one for the maximum number of period
- fixed the typo in comment and updated the commit message

Note: When sending for upstream, we should mark it for stable backport, I'll test the oldest I can find where it applies.

ujfalusi commented:
Interesting, legacy HDA (and AVS, skl) is using 32 as periods_max (AZX_MAX_FRAG)...

plbossart commented:
> Interesting, legacy HDA (and AVS, skl) is using 32 as periods_max (AZX_MAX_FRAG)...

hda_register.h:#define BDL_SIZE         4096
hda_register.h:#define AZX_MAX_BDL_ENTRIES      (BDL_SIZE / 16)

plbossart commented:
I was trying to find the explanation for this comment 

````c
/* max number of fragments - we may use more if allocating more pages for BDL */
````

That comment is older than 2.6.12-rc2 and git history. We may need to ask @tiwai why there's a restriction to 32 periods?

ujfalusi commented:
> I was trying to find the explanation for this comment
> 
> ```c
> /* max number of fragments - we may use more if allocating more pages for BDL */
> ```
> 
> That comment is older than 2.6.12-rc2 and git history. We may need to ask @tiwai why there's a restriction to 32 periods?

Headroom to handle non continuous ALSA buffers perhaps?
